### PR TITLE
Reset grad attribute when called using del

### PR DESCRIPTION
**Pull request by bhushan23**
del Tensor.grad set PyObject to nullptr
and Tensor.grad = None set PyObject to Py_None
Handling both the cases now
fixes ##16471

### CODE REVIEW

**Comment by bhushan23** [comment]
@colesbury as we cannot control what PyObject gets, we need to add above check.
Is it correct? If so, Will add respective test cases

**Comment by colesbury** [comment]
Yes. This same problem occurs in some other setters in THPVariable_properties and maybe other classes. Including:

- THPVariable_set_grad_fn
- THPVariable_set_data
- THPVariable_set_requires_grad

(i.e. they crash when the someone tries to delete the property)

**Comment by vishwakftw** [comment]
@colesbury Is `grad_fn` deletable? `grad_fn` is not writable, I believe.

https://github.com/pytorch/pytorch/blob/21193bf12314f9d1be5ea36cbce3162c54b35225/torch/csrc/autograd/python_variable.cpp#L409

File: test/test_autograd.py [hunk 1]
```diff
@@ -1093,6 +1093,15 @@ def test_requires_grad_factory(self):
                             self.assertIs(dtype, output.dtype)
                             self.assertEqual(1, output.get_device())
 
+    def test_attribute_deletion(self):                                                            
+        x = torch.randn((5,5), requires_grad = True)
```
Comment (vishwakftw) [inline test/test_autograd.py]: Spaces before and after `=` would probably cause lint error, and so will no space after `,` in `(5,5)`.

**Comment by bhushan23** [comment]
@colesbury

File: test/test_autograd.py [hunk 2]
```diff
@@ -1093,6 +1093,15 @@ def test_requires_grad_factory(self):
                             self.assertIs(dtype, output.dtype)
                             self.assertEqual(1, output.get_device())
 
+    def test_attribute_deletion(self):
+        x = torch.randn((5, 5), requires_grad=True)
+        del x.grad
+        self.assertIsNone(x.grad)
+        del x.data
+        self.assertIsNone(x.data)
```
Comment (albanD) [inline test/test_autograd.py]: What does it mean for a Tensor `.data` field to be `None` ? What happens to `x` when you do `del x.data` ? Does it become unusable?

Comment (albanD) [reply test/test_autograd.py]: From my point of view, not allowing it would be better as this is a state you never want to be in (in python).

**Comment by bhushan23** [comment]
@colesbury

**Review by fmassa — APPROVED** [review]
I think this looks good, but I'll let @colesbury have another look at it

**Comment by soumith** [comment]
just fyi to all, @colesbury is on vacation. If @albanD also approves, I'm happy to ship it.

**Review by albanD — CHANGES_REQUESTED** [review]
Could you do the same for `THPVariable_set_grad_fn` and forbid deleting it please (A user should never delete a grad_fn but detach() the Tensor) and `THPVariable_set_backwards_hooks` to make it have the same behavior as if None was passed (this shuold not be used by users, but I think it's better to remove the case where it segfaults)? Both these functions have the same problem as the one you fixed.

**Comment by bhushan23** [comment]
@albanD sounds good to me. 
Unfortunately, I do not have my laptop with me and can not setup pytorch on temporary laptop. Will send change in next couple of days.

**Comment by bhushan23** [comment]
@albanD not allowing deletion for `THPVariable_set_grad_fn`
`THPVariable_set_backwards_hooks`'s behavior for deletion and setting to None is same, hence skipping.

**Comment by albanD** [comment]
But won't the check `obj == Py_None` segfault if `obj` is already a `nullptr` for `THPVariable_set_backwards_hooks`?

**Comment by bhushan23** [comment]
> But won't the check `obj == Py_None` segfault if `obj` is already a `nullptr` for `THPVariable_set_backwards_hooks`?
@albanD No.  It won't as obj pointer is not same as one pointing to Py_None

But, if we don't want to allow deletion of backward hooks, we should definitely assert with useful error message.

**Comment by albanD** [comment]
Ho my bad, for some reason I though it was the equality check that was causing issues. We are happy with deletion to behave like setting it to None !

**Review by albanD — APPROVED** [review]

**Comment by bhushan23** [comment]
> Ho my bad, for some reason I though it was the equality check that was causing issues. We are happy with deletion to behave like setting it to None !

FYI, I just pushed changes to assert if del is called on _backward_hooks
Fact that, deletion and setting to None behavior is same is good to have but do we want to allow it? 
Not allowing will be good as we keep consistent behavior with other grad attribute

**Comment by albanD** [comment]
This particular method is an internal one. It is just used internally at the moment.
Also you don't want the user to tinker with grad_fn (there are proper functions for that) but you're happy with him removing the hooks.
